File: sandboxes/js/sandbox.ts
import { task } from "@capsule-run/sdk";
import path from "path";
import fs from "fs";

import type { State } from "@capsule-run/bash-types";

function wasmRelative(cwd: string, filePath: string): string {
  return path.resolve(cwd, filePath).replace(/^\//, '');
}

const executeFile = task(
  { name: "executeFile", compute: "MEDIUM", ram: "512MB" },
  async (state: State, filePath: string, args: string[]) => {
    const capturedOutput: string[] = [];
    const relPath = wasmRelative(state.cwd, filePath);

    Object.defineProperty(process, 'argv', {
      value: ['node', relPath, ...args],
      writable: true,
      configurable: true,
    });

    const originalLog = console.log;
    console.log = (...logArgs: any[]) => {
      capturedOutput.push(logArgs.map(a => String(a)).join(' '));
    };

    try {
      const code = fs.readFileSync(relPath, 'utf-8') as string;
      const mod = { exports: {} as any };

      const makeRequire = (fromPath: string) => (id: string) => {
        const base = wasmRelative('/', path.resolve('/' + path.dirname(fromPath), id));
        const depPath = fs.existsSync(base) ? base
          : fs.existsSync(base + '.js') ? base + '.js'
          : base + '/index.js';
        const src = fs.readFileSync(depPath, 'utf-8') as string;
        const m = { exports: {} as any };
        const fn = new Function('module', 'exports', 'require', '__filename', '__dirname', src);
        fn(m, m.exports, makeRequire(depPath), depPath, path.dirname(depPath));
        return m.exports;
      };
      const customRequire = makeRequire(relPath);

      const fn = new Function('module', 'exports', 'require', '__filename', '__dirname', code);
      fn(mod, mod.exports, customRequire, relPath, path.dirname(relPath));

      const output = capturedOutput.join('\n');
      if (output) {
        return Object.keys(mod.exports).length > 0
          ? output + '\n' + JSON.stringify(mod.exports)
          : output;
      }
      return mod.exports;
    } finally {
      console.log = originalLog;
    }
  }
);


const executeCode = task(
  { name: "executeCode", compute: "LOW", ram: "256MB" },
  async (state: State, code: string): Promise<unknown> => {
    process.chdir(state.cwd);
    const capturedOutput: string[] = [];
    const originalLog = console.log;

    console.log = (...args: any[]) => {
      capturedOutput.push(args.map(arg => String(arg)).join(' '));
    };

    try {
      let result;
      try {
        result = eval(code);
      } catch (e) {
        if (e instanceof SyntaxError && e.message.includes("return")) {
          const fn = new Function(code);
          result = fn();
        } else {
          throw e;
        }
      }

      const output = capturedOutput.join('\n');

      if (output) {
        return output + '\n' + result;
      }

      return result;
    } finally {
      console.log = originalLog;
    }
  }
);

const executeCommand = task(
  { name: "executeCommand", compute: "LOW", ram: "64MB" },
  async (state: State, scriptContent: string, args: string[]) => {
    process.chdir(state.cwd);
    const exports: { execute?: (args: string[]) => any } = {};

    console.log(args)
    const moduleWrapper = new Function('exports', scriptContent);

    moduleWrapper(exports);

    if (!exports.execute) {
      throw new Error("Script must export an 'execute' function");
    }

    return await exports.execute(args);
  }
)

export const resolveDirectoryPath = task(
  { name: "resolveDirectoryPath", compute: "LOW", ram: "32MB" },
  async (state: State, targetPath: string) => {
    process.chdir(state.cwd);

    if (!fs.existsSync(targetPath)) {
      throw new Error(`Directory path ${targetPath} does not exist`);
    }

    return '/' + path.resolve(targetPath);
  }
)

export const main = task(
  { name: "main", compute: "HIGH" },
  async (action: string, state: string, ...args: string[]): Promise<unknown> => {
    let parsedArgs = args.filter(arg => typeof arg !== "object"); // remove the unused kwargs
    let response: { success: boolean; result: unknown; error: { error_type: string; message: string } | null };

    let parsedState: State = JSON.parse(state);

    if (action === "LOAD") {
      response = { success: true, result: "Sandbox loaded successfully", error: null };
    } else if (action === "EXECUTE_COMMAND") {
      response = await executeCommand(parsedState, parsedArgs[0], parsedArgs.slice(1));
    } else if (action === "EXECUTE_CODE") {
      response = await executeCode(parsedState, parsedArgs[0]);
    } else if (action === "EXECUTE_FILE") {
      response = await executeFile(parsedState, parsedArgs[0], parsedArgs.slice(1));
    } else if (action === "RESOLVE_DIRECTORY_PATH") {
      response = await resolveDirectoryPath(parsedState, parsedArgs[0]);
    } else {
      throw new Error(`Invalid action: ${action}`);
    }

    if (!response.success && response.error) {
      throw new Error(response.error.message);
    }

    return response.result;
  }
);

File: sandboxes/python/sandbox.py
import ast
import sys
import os
import json
import runpy
from io import StringIO
from dataclasses import dataclass
from capsule import task


@dataclass
class State:
    cwd: str
    env: dict
    lastExitCode: int

    @staticmethod
    def from_json(raw: str) -> "State":
        data = json.loads(raw)
        return State(
            cwd=data["cwd"],
            env=data["env"],
            lastExitCode=data["lastExitCode"],
        )


@task(name="executeFile", compute="LOW", ram="256MB")
def execute_file(state: State, file_path: str, args: list[str]):
    os.chdir(state.cwd)

    absolute_path = os.path.realpath(os.path.join(state.cwd, file_path))

    captured_output = StringIO()
    old_stdout = sys.stdout
    old_argv = sys.argv
    sys.stdout = captured_output
    sys.argv = [absolute_path] + list(args)

    try:
        result = runpy.run_path(absolute_path, run_name="__main__")
    finally:
        sys.stdout = old_stdout
        sys.argv = old_argv

    output = captured_output.getvalue()
    public_result = {k: v for k, v in result.items() if not k.startswith("__")}

    if output:
        return output.rstrip("\n") + "\n" + json.dumps(public_result) if public_result else output.rstrip("\n")

    return public_result if public_result else None


@task(name="executeCode", compute="LOW", ram="256MB")
def execute_code(state: State, code: str):
    os.chdir(state.cwd)

    tree = ast.parse(code)

    if not tree.body:
        return None

    last_node = tree.body[-1]
    local_env = {}

    captured_output = StringIO()
    old_stdout = sys.stdout
    sys.stdout = captured_output

    try:
        if isinstance(last_node, ast.Expr):
            tree.body.pop()
            if tree.body:
                exec(compile(tree, filename="<ast>", mode="exec"), local_env)
            result = eval(
                compile(ast.Expression(last_node.value), filename="<ast>", mode="eval"),
                local_env,
            )
        else:
            exec(compile(tree, filename="<ast>", mode="exec"), local_env)
            result = local_env.get("result")
    finally:
        sys.stdout = old_stdout

    output = captured_output.getvalue()

    if output:
        return output + str(result)

    return result


@task(name="executeCommand", compute="LOW", ram="64MB")
def execute_command(state: State, script_content: str, args: list[str]):
    os.chdir(state.cwd)

    module_env = {}
    exec(compile(script_content, filename="<script>", mode="exec"), module_env)

    execute_fn = module_env.get("execute")
    if not callable(execute_fn):
        raise ValueError("Script must define an 'execute' function")

    return execute_fn(list(args))


@task(name="main", compute="HIGH")
def main(action: str, state: str, *args: str):
    parsed_state = State.from_json(state)

    if action == "LOAD":
        response = {"success": True, "result": "Sandbox loaded successfully", "error": None}
    elif action == "EXECUTE_COMMAND":
        response = execute_command(parsed_state, args[0], list(args[1:]))
    elif action == "EXECUTE_CODE":
        response = execute_code(parsed_state, args[0])
    elif action == "EXECUTE_FILE":
        response = execute_file(parsed_state, args[0], list(args[1:]))
    else:
        raise ValueError(f"Invalid action: {action}")

    if isinstance(response, dict):
        if not response.get("success"):
            raise Exception(response["error"]["message"])
        if response.get("result") is not None:
            return response["result"]

    return response
